core/src/main/java/org/apache/iceberg/hadoop/HadoopTableOperations.java

```diff
@@ -289,19 +311,24 @@ int readVersionHint() {
       return Integer.parseInt(in.readLine().replace("\n", ""));
 
     } catch (Exception e) {
-      LOG.warn("Error reading version hint file {}", versionHintFile, e);
+      LOG.debug("Error reading version hint file {}", versionHintFile, e);
```

Comment:
Why make this debug instead of warn? I think it is still concerning that the version hint couldn't be found.

Reply:
I'd go one step further. I think a user should have to opt into this fallback behavior. I think the default should be to throw an exception. Silently "corrupting" the table isn't something that should happen.

Reply:
The warn message appeared several times for me and I have realized that we call this method even for new tables to check that the data/metadata is not yet there, so this warning message could be misleading.

@jacques-n: Adding a configuration seems like a good idea because this recovery mechanisms starts to become complicated. Could you help by pointing out some similar configurable solutions from where I can start from?

Thanks,
Peter

Reply:
@rdblue we've seen this happening with hadoop implementations that implement `create(overwrite=true)` by a server-side `delete` and `create` operation - so for a split micro-second we get a 404 on the version-hint.text file - we've looked at solving this in a pretty similar file but instead of a fallback mechanism we went with the approach that a table configuration (or version?) indicates the version resolver implementation, either file based or directory listing based.

@jacques-n I don't think that this approach to fallback to directory listing in case of failing to resolve the version hint file can lead to silent "corruption" of the table. Can you think of such a scenario? I'm asking cause we were considering doing something similar so I'm interested in any thoughts that may question this approach.
But to your point if there so much is a possibility that this approach can lead to table corruption then we should probably not provide this feature at all, let alone say support it by a "turn-this-feature-at-own-risk" approach - it will still corrupt the table :)

@pvary Big difference in the listing directory vs file approach is the constant time in loading data from a file (constant time) vs using hadoop listing directory (probably not constant time).
I would assume it would take longer to resolve the version and probably it would be advised that older versions are dropped so the list API provide a decent latency.

Reply:
>  I think this approach is actually better cause it's simple however it doesn't control the latency of the list API...

If I understand correctly in HadoopTableOperations.refresh() we use the number provided by versionHint() only as a basis for checking newer versions. Would that cover your concerns?
```
      Path nextMetadataFile = getMetadataFile(ver + 1);
      while (nextMetadataFile != null) {
        ver += 1;
        metadataFile = nextMetadataFile;
        nextMetadataFile = getMetadataFile(ver + 1);
      }

      updateVersionAndMetadata(ver, metadataFile.toString());
```

> @pvary Big difference in the listing directory vs file approach is the constant time in loading data from a file (constant time) vs using hadoop listing directory (probably not constant time).

With the current patch we still do the fast / file based version resolving, and only fall back to the listing based version if there is a problem with the version-hint.txt file.

> I would assume it would take longer to resolve the version and probably it would be advised that older versions are dropped so the list API provides a decent latency.

Based on my current understanding the older versions might be needed for timetravel or other features. It should be really the decision of the users to decide when they want to remove them.

Reply:
I think @jacques-n has a good point. The version file should not be corrupt and we should make sure of it by changing how we write the file. Since this is for HDFS, creating the file with an atomic rename to ensure the entire file is written before making it the current version hint makes sense to me. That way we don't get dirty reads.

Doing that would result in cases like @fbocse brings up: we'd have to delete the version hint file to rename into its place. And there are cases where the hint file is missing, like when trying to load a table to check whether it exists. In that case, I don't think that it makes sense to warn when a hint file is missing, only when it is corrupt. It would still be good to have an info log when the file is missing and there is no metadata (table doesn't exist), and a warning if the file is missing but a metadata file is found.

Last, I'm for an option to throw an exception when the hint file is corrupt rather than throwing a warning, but the hint file is not intended as a requirement for correctness, so I'd turn this off by default.

Reply:
Looking at the code, I was wrong about the table corruption. I thought that the version hint file was used in a slightly different manner. I agree with @rdblue that the default probably shouldn't be exception.

Reply:
> I think @jacques-n has a good point. The version file should not be corrupt and we should make sure of it by changing how we write the file. Since this is for HDFS, creating the file with an atomic rename to ensure the entire file is written before making it the current version hint makes sense to me. That way we don't get dirty reads.

I think that is something for another PR.

> It would still be good to have an info log when the file is missing and there is no metadata (table doesn't exist), and a warning if the file is missing but a metadata file is found.

Here is my proposed solution:
- No metadata directory: DEBUG log - there are some perfectly valid cases where we expect that the directory/table is not there. Higher level log message could be very confusing (Seen very bad examples in Hive code 😢)
- Existing metadata directory, no hint file: WARN log - transient error, or hint file error. Good to be aware of.
- IOException while checking for existence of the metadata directory or listing: WARN log - Some fs level exception. Has to be checked.

Your thoughts?

> Last, I'm for an option to throw an exception when the hint file is corrupt rather than throwing a warning, but the hint file is not intended as a requirement for correctness, so I'd turn this off by default.

Could someone please point me a place (code lines) where and how Iceberg configurations are handled, so I can use them correctly?

Thanks,
Peter

Reply:
> Could someone please point me a place (code lines) where and how Iceberg configurations are handled, so I can use them correctly?

I don't think we need this option since Jacques replied that he agrees it should not be an exception.

For configuration, we prefer to use table properties for most configuration, which are defined in the `TableProperties` class. For other cases, we use the configuration that makes the most sense. Here, I would use the Hadoop configuration because the table implementation is for Hadoop and HDFS.